src/main/scala/uclid/ConcreteSimulator.scala

```
@@ -0,0 +1,1293 @@
+                            simulate_stmt(context,body,iter,module)
+                        }
+                    }
+                    case _ => throw new Error("Does not support loop index of type "+ typ.toString)
+                }
+                context.removeVar(List((id,typ)))}
+            case WhileStmt(cond, body, invariants) => {
+                while(evaluateBoolExpr(context, cond)){
+                    simulate_stmt(context, body,iter,module)
+                }}
+            case CaseStmt(body) => throw new NotImplementedError("We have not implemented Case stmt")
+            case ProcedureCallStmt(id, callLhss, args, instanceId, moduleId) => {
+                throw new NotImplementedError(s"ProcedureCallStmt not implemented")}
+            case MacroCallStmt(id) => {
+                throw new NotImplementedError(s"MacroCallStmt not implemented")}
+            case ModuleCallStmt(id) => {
+                throw new NotImplementedError(s"ModuleCallStmt not implemented")}
+            case _ => throw new NotImplementedError("We have not implemented Stmt "+ stmt.toString)
+        }}
+    def evaluateBoolExpr(context: ConcreteContext,
+        cond: Expr) : Boolean = {
+            printDebug("Evaluate BoolExpr "+cond.toString)
+            evaluate_expr(context,cond) match {
+                case ConcreteBool(b) => {
+                    if (b) {
+                        return true
+                    } else {
+                        return false
+                    }
+                }
+                //TODO:
+                //make better check on this part
+                case ConcreteUndef() => {
+                    //if we hit undef value, just give a false
+                    //println("Hit a Undef Boolean")
+                    true
+                    // context.printVar(List())
+                    // throw new Error("When Evaluation Bool value we hit a undefine value "+cond.toString)
+                }
```

again seems like you should throw an error here?